compiler/hash-ir/src/constant.rs

```diff
@@ -0,0 +1,537 @@
+        } else {
+            None
+        }
+    }
+
+    /// Convert an unsigned integer value into a [Scalar].
+    #[inline]
+    pub fn from_int(i: impl Into<i128>, size: Size) -> Self {
+        Self::try_from_int(i, size).unwrap_or_else(|| {
+            panic!("Integer value is too large for the specified size: {}", size.bytes())
+        })
+    }
+
+    /// Create a new [Scalar] from a boolean value.
+    #[inline]
+    pub fn from_bool(value: bool) -> Self {
+        if value {
+            Self::TRUE
+        } else {
+            Self::FALSE
+        }
+    }
+
+    /// Convert the [Scalar] into a [f32].
+    pub fn to_f32(&self) -> f32 {
+        f32::try_from(*self).unwrap()
+    }
+
+    /// Convert the [Scalar] into a [f64].
+    pub fn to_f64(&self) -> f64 {
+        f64::try_from(*self).unwrap()
+    }
+}
+
+impl From<bool> for Scalar {
+    fn from(value: bool) -> Self {
+        Self {
+            value: value as u128,
+            size: NonZeroU8::new(std::mem::size_of::<bool>() as u8).unwrap(),
+        }
+    }
+}
+
+impl TryFrom<Scalar> for bool {
+    type Error = Size;
+    #[inline]
+    fn try_from(value: Scalar) -> Result<Self, Self::Error> {
+        value.to_bits(Size::from_bytes(1)).and_then(|u| match u {
+            0 => Ok(false),
+            1 => Ok(true),
+            _ => Err(Size::from_bytes(1)),
+        })
+    }
+}
+
+impl From<char> for Scalar {
+    fn from(value: char) -> Self {
+        Self {
+            value: value as u128,
+            size: NonZeroU8::new(std::mem::size_of::<char>() as u8).unwrap(),
+        }
+    }
+}
+
+impl TryFrom<Scalar> for char {
+    type Error = ();
+
+    fn try_from(value: Scalar) -> Result<Self, Self::Error> {
+        let Ok(val) = value.to_bits(Size::from_bytes(std::mem::size_of::<char>())) else {
+            return Err(());
+        };
+
+        match char::from_u32(val.try_into().unwrap()) {
+            Some(c) => Ok(c),
+            None => Err(()),
+        }
+    }
+}
+
+impl From<IntConstant> for Scalar {
+    fn from(value: IntConstant) -> Self {
+        let size = value.size();
+
+        match value.value {
+            IntConstantValue::I8(val) => Scalar::from_int(val, size),
+            IntConstantValue::I16(val) => Scalar::from_int(val, size),
+            IntConstantValue::I32(val) => Scalar::from_int(val, size),
+            IntConstantValue::I64(val) => Scalar::from_int(val, size),
+            IntConstantValue::I128(val) => Scalar::from_int(val, size),
+            IntConstantValue::U8(val) => Scalar::from_uint(val, size),
+            IntConstantValue::U16(val) => Scalar::from_uint(val, size),
+            IntConstantValue::U32(val) => Scalar::from_uint(val, size),
+            IntConstantValue::U64(val) => Scalar::from_uint(val, size),
+            IntConstantValue::U128(val) => Scalar::from_uint(val, size),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Constructor)]
+pub struct ScalarInt {
```

Comment:
What is the purpose of this struct? Is it to "parse" a scalar into an int?

Reply:
Its only really used for actually printing the `Scalar` since the scalar itself has no type info on it.